api/handlers_servers.go

```diff
@@ -0,0 +1,215 @@
+package api
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/YaleSpinup/ds-api/apierror"
+	"github.com/YaleSpinup/ds-api/dataset"
+	"github.com/gorilla/mux"
+	log "github.com/sirupsen/logrus"
+)
+
+// ServerCreateHandler provisions access to a "dataset" for a given server
+func (s *server) ServerCreateHandler(w http.ResponseWriter, r *http.Request) {
+	w = LogWriter{w}
+	vars := mux.Vars(r)
+	account := vars["account"]
+	id := vars["id"]
+
+	service, ok := s.datasetServices[account]
+	if !ok {
+		msg := fmt.Sprintf("account not found: %s", account)
+		handleError(w, apierror.New(apierror.ErrNotFound, msg, nil))
+		return
+	}
+
+	input := struct {
+		InstanceID string `json:"instance_id"`
+	}{}
+
+	err := json.NewDecoder(r.Body).Decode(&input)
+	if err != nil {
+		msg := fmt.Sprintf("cannot decode body into create server input: %s", err)
+		handleError(w, apierror.New(apierror.ErrBadRequest, msg, err))
+		return
+	}
+
+	if input.InstanceID == "" {
+		handleError(w, apierror.New(apierror.ErrBadRequest, "instance_id is required", nil))
+		return
+	}
+
+	log.Infof("provisioning access to data set '%s' in account '%s' for server: %s", id, account, input.InstanceID)
+
+	metadataOutput, err := service.MetadataRepository.Get(r.Context(), account, id)
+	if err != nil {
+		handleError(w, err)
+		return
+	}
+
+	dataRepo, ok := service.DataRepository[metadataOutput.DataStorage]
+	if !ok {
+		msg := fmt.Sprintf("requested data repository type not supported for this account: %s", metadataOutput.DataStorage)
+		handleError(w, apierror.New(apierror.ErrBadRequest, msg, nil))
+		return
+	}
+
+	// list current access to this data repository
+	listAccess, err := dataRepo.ListAccess(r.Context(), id)
+	if err != nil {
+		msg := fmt.Sprintf("failed to list access to data repository for dataset %s: %s", id, err)
+		handleError(w, apierror.New(apierror.ErrInternalError, msg, err))
+		return
+	}
+
+	// check if requested server already has access
+	if _, found := listAccess[input.InstanceID]; found {
```

Comment:
If we don't do this check it will continue to `GrantAccess` and re-attach the policy, which doesn't really hurt, just depends if we want to return a `409` in this case or a `200`

Reply:
maybe thats okay?  if we think it's safe that is.  that listAccess does a bunch of stuff and makes a bunch of calls to AWS.